Remove usage of "run_until_complete"

smurfix opened the pull request:
## Description:

This patch changes Home Assistant's startup routine so that it will run under `asyncio.run`. It also changes the SIGINT handler to start the shutdown sequence.

Advantages:
* This allows people to embed Home Assistant in larger async programs.
* SIGINT aka ^C no longer disrupts a random part of Home Assistant's code.

Disadvantages:
* If Home Assistant's main thread is wedged (which basically should not happen), you may have to press ^C twice to get out.

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

smurfix committed:
De-run_forever()-ization

* Use asyncio.run (or our own implementation on Python <3.7)
* hass.start is only used by tests
* setup_and_run_hass() is now async
* Add "main" async hass.run method
* move SIGINT handling to helpers/signal.py
  * add flag to .run to disable hass's signal handlers
* Teach async_start and async_stop to not step on each other
  (more than necessary)

smurfix committed:
shorten over-long lines

balloob commented on homeassistant/__main__.py:
```diff
@@ -21,8 +21,6 @@
 
 def attempt_use_uvloop() -> None:
     """Attempt to use uvloop."""
-    import asyncio
```
Now the import is missing?

smurfix replied:
Ugh. Me being blind. Will fix.

balloob commented on homeassistant/core.py:
```diff
@@ -321,13 +351,24 @@ def block_till_done(self) -> None:
+        if self.state == CoreState.not_running:  # just ignore
+            return
+        if self.state == CoreState.stopping:
+            _LOGGER.info("async_stop called twice: ignored")
```
Should this be a warning?

smurfix replied:
Might be a better idea. Will change.

balloob commented on homeassistant/util/__init__.py:
```diff
@@ -30,6 +30,15 @@
 }
 
 
+try:
```
Can we move this to `util.async_`

balloob commented on homeassistant/helpers/signal.py:
```diff
@@ -17,7 +17,12 @@ def async_register_signal_handling(hass: HomeAssistant) -> None:
+            * queue shutdown code
+            * re-instate default handler
+            """
+            hass.loop.remove_signal_handler(signal.SIGTERM)
```
Why would we do this ?

smurfix replied:
Because the signal handler has initiated an orderly shutdown. Its job is now done; if we get the signal again, the user is impatient / the main loop is stuck, and standard Python signal handling should take over.

smurfix committed:
restore missing "import asyncio"

smurfix committed:
move run_asyncio to homeassistant.util.async_

smurfix committed:
LOGGER: warn => warning

smurfix committed:
Add "force" flag to async_stop

only useful for testing

smurfix committed:
Add 'attrs==18.2.0' to requirements_all.txt

Required for keeping requirements_test_all.txt in sync, where it is in
turn required to prevent auto-downgrading "attrs" during "pip install"

smurfix committed:
Fixes for mypy

smurfix commented:
Current tests break with undefined "mock_signal" fixture.
I have no idea where that is supposed to come from …

smurfix committed:
Fix "mock_signal" fixture

smurfix committed:
Revert mistaken edit

smurfix committed:
Flake8 fixes

smurfix committed:
mypy fixes

smurfix committed:
pylint fix

smurfix commented:
Meh. Finally. (Had a broken local installation.)
Ready to merge IMHO.

balloob commented on homeassistant/core.py:
```diff
@@ -154,30 +154,54 @@ def __init__(
             self.loop.close()
         return self.exit_code
 
+    async def run(self, *, attach_signals: bool = True) -> int:
```
If this is an async method, we should call it `async_run` to be in line with the rest of our functions.

smurfix replied:
My rationale for not naming it `async_run` is that it's never called from within Home Assistant, but I'll change it if you want me to, no problem.

balloob replied:
Yeah let's change it.

balloob commented on requirements_test.txt:
```diff
@@ -2,6 +2,7 @@
 # make new things fail. Manually update these pins when pulling in a
 # new version
 asynctest==0.12.2
+attrs==18.2.0
```
This shouldn't be necessary, it's already part of the core dependencies.

smurfix replied:
The problem: when I do a `pip install -r requirements_test_all.txt` without this entry, `attrs` is downgraded to some lower version, probably because of some sub-module's requirements.

smurfix committed:
Revert adding attrs== to requirements_test*.txt

solved by using "pip -c"

balloob reviewed (approved):
ok to merge after run renamed to async_run

smurfix committed:
Rename "run" to "async_run", as per calling conventions

smurfix commented:
All done, please merge.

awarecan commented:
Did it tested on Windows?

I run on Windows Python 3.6.3 got some errors
```
Traceback (most recent call last):
  File "c:\awarecan\home-assistant\homeassistant\__main__.py", line 376, in <module>
    sys.exit(main())
  File "c:\awarecan\home-assistant\homeassistant\__main__.py", line 368, in main
    exit_code = asyncio_run(setup_and_run_hass(config_dir, args))
  File "c:\awarecan\home-assistant\homeassistant\util\async_.py", line 29, in asyncio_run
    return loop.run_until_complete(main)
  File "C:\Python\Python36\lib\asyncio\base_events.py", line 467, in run_until_complete
    return future.result()
  File "c:\awarecan\home-assistant\homeassistant\__main__.py", line 275, in setup_and_run_hass
    log_no_color=args.log_no_color)
  File "c:\awarecan\home-assistant\homeassistant\bootstrap.py", line 215, in async_from_config_file
    conf_util.load_yaml_config_file, config_path)
RuntimeError: Task <Task pending coro=<setup_and_run_hass() running at c:\awarecan\home-assistant\homeassistant\__main__.py:275> cb=[_run_until_complete_cb() at C:\Python\Python36\lib\asyncio\base_events.py:176]> got Future <Future pending cb=[_chain_future.<locals>._call_check_cancel() at C:\Python\Python36\lib\asyncio\futures.py:408]> attached to a different loop
```

balloob commented:
This logic should be moved to main to fix it:

https://github.com/home-assistant/home-assistant/blob/fcb84d951e8022ace87bb19be259f401c9920ddc/homeassistant/core.py#L132-L135

balloob commented:
This should do it: https://github.com/home-assistant/home-assistant/pull/16737
